src/libworld/libworld.h

```diff
@@ -2549,137 +2553,103 @@ namespace world
         {
-            m_rng = randomGenerator;
+            const vector<shared_ptr<Route>> & routesTo = getValueOrThrow(m_routesTo, toIcao);
+            return getNextRoute(routesTo, m_routesToIter[toIcao]);
         }
-        const Route &findRandomRouteFrom(const string &fromICAO, const string &airframe, const vector<string> &allowedAirlines)
+        size_t routesFromCount(const string &fromIcao)
         {
-            return findRandomRoute(getValueOrThrow(m_routesFrom, fromICAO), airframe, allowedAirlines);
+            return m_routesFrom[fromIcao].size();
         }
-        const Route &findRandomRouteTo(const string &toICAO, const string &airframe, const vector<string> &allowedAirlines)
+        size_t routesToCount(const string &toIcao)
         {
-            return findRandomRoute(getValueOrThrow(m_routesTo, toICAO), airframe, allowedAirlines);
+            return m_routesTo[toIcao].size();
         }
 
     private:
-        const Route &findRandomRoute(const vector<shared_ptr<world::RouteProvider::Route>> _routes, const string &airframe, const vector<string> &allowedAirlines)
+        const Route &getNextRoute(const vector<shared_ptr<Route>> &routes, vector<shared_ptr<Route>>::const_iterator &iter)
         {
-            vector<shared_ptr<world::RouteProvider::Route>> routes(_routes);
-            shuffle(routes.begin(), routes.end(), m_rng);
-
-            do
+            if (iter == routes.end())
             {
-                auto route = routes.back();
-                routes.pop_back();
-                auto allowedAirFrames = route->usedAirframes();
-                // Any airline is allowed if allowedAirlines is empty
-                auto routeAirline = (allowedAirlines.size() > 0) ? route->airline() : "";
-
-                // Airframe asked must be used on this route
-                if (!airframe.empty() && !hasStringInsensitive(allowedAirFrames, airframe))
-                {
-                    continue;
-                }
-
-                // Route must be operated by one of the airlines allowed
-                if (!allowedAirlines.empty() && !hasStringInsensitive(allowedAirlines, routeAirline))
-                {
-                    continue;
-                }
-                // return found route
-                return *route;
-            } while (!routes.empty());
-
-            // No route found
-            throw std::runtime_error("Route not found");
-        }
+                // we could get here if there are no more routes in the vector.
 
-        static bool iequals(const string &a, const string &b)
-        {
-            if (a.length() == b.length())
+                // The other possibility is that someoine played with the iterator
+                throw(runtime_error("No more routes or bad iterator usage"));
+            }
+            const Route &result = **iter;
+            iter++;
+            // This should not be allowed when routes have a schedule
```

Comment:
> This should not be allowed when routes have a schedule

Why?

Reply:
If/when we'll have routes with departure/arrival times, for example :

- Route 1
   - from : KJFK
   - to: KMIA
   - departure 2020/11/25 09:30
   - arrival 2020/11/25 12:30
- Route 2
   - from : KJFK
   - to: KDEN
   - departure 2020/11/25 10:30
   - arrival 2020/11/25 14:30

We probably should stop providing routes when all the routes are assigned a pilot/aircraft and not give the same route/planning to multiple pilot/aircrafts.

Reply:
Agree